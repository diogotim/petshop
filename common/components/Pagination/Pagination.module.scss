.root {
    display: flex;
    justify-content: center;
    margin: 2rem 0;
    gap: 4rem;
    font-family: var(--display-font);
    font-size: 1.2em;

    a:hover {
        color: var(--main);
    }

    .active {
        color: var(--main);
    }
}