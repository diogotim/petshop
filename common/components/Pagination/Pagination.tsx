import React from "react";
import Link from "next/link";
import classes from "./Pagination.module.scss";
import { Product } from "../../../types";
import { arrayFromRange } from "../../util/util";

type Props = {
  perPage: number;
  currentPage: number;
  total: number;
  prefix: string;
};

const MAX_PAGES_SHOWN = 6;

const getPageButtons = (
  pages: number[],
  currentPage: number,
  prefix: string
) => {
  return pages.map((p) => (
    <Link href={`${prefix}${p}`} key={p}>
      <a className={currentPage === p ? classes.active : ""}>{p}</a>
    </Link>
  ));
};

const getPages = ({ perPage, currentPage, total, prefix }: Props) => {
  if (total <= MAX_PAGES_SHOWN) {
    return getPageButtons(arrayFromRange(1, total), currentPage, prefix);
  }

  if (currentPage < 3) {
    return getPageButtons(
      arrayFromRange(1, MAX_PAGES_SHOWN),
      currentPage,
      prefix
    );
  }

  if (currentPage > total - MAX_PAGES_SHOWN) {
    return getPageButtons(
      arrayFromRange(total - MAX_PAGES_SHOWN, total),
      currentPage,
      prefix
    );
  }

  return getPageButtons(
    [
      ...arrayFromRange(currentPage - 3, currentPage),
      ...arrayFromRange(currentPage + 1, currentPage + 3),
    ],
    currentPage,
    prefix
  );
};

const Pagination: React.FunctionComponent<Props> = ({
  perPage,
  currentPage,
  total,
  prefix,
}) => {
  return (
    <div className={classes.root}>
      {getPages({ perPage, currentPage, total, prefix })}
    </div>
  );
};

export default Pagination;
