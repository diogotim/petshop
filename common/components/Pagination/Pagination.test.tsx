import { render } from "@testing-library/react";
import Pagination from "./Pagination";

describe("Pagination", () => {
  it("renders the pagination unchanged", () => {
    const pag = render(
      <Pagination perPage={6} currentPage={1} total={12} prefix="/products/" />
    );
    expect(pag).toMatchSnapshot();
  });

  it("renders all pages if there's lesser than the maximum", () => {
    const { getAllByRole } = render(
      <Pagination perPage={6} currentPage={1} total={4} prefix="/products/" />
    );
    expect(getAllByRole("link").length).toEqual(4);
  });

  it("do not render more than the page limit", () => {
    const { getAllByRole } = render(
      <Pagination perPage={6} currentPage={1} total={20} prefix="/products/" />
    );
    expect(getAllByRole("link").length).toEqual(6);
  });

  it("renders starting pages if the current page is low", () => {
    const { getByText } = render(
      <Pagination perPage={6} currentPage={1} total={12} prefix="/products/" />
    );
    for (let i = 1; i < 7; i += 1) {
      expect(getByText(i.toString())).toBeTruthy();
    }
  });

  it("renders last pages if the current page is high", () => {
    const { getByText } = render(
      <Pagination perPage={6} currentPage={10} total={12} prefix="/products/" />
    );
    for (let i = 7; i < 13; i += 1) {
      expect(getByText(i.toString())).toBeTruthy();
    }
  });

  it("renders middle pages if the current page is not low or high", () => {
    const { getByText } = render(
      <Pagination perPage={6} currentPage={5} total={12} prefix="/products/" />
    );
    for (let i = 3; i < 9; i += 1) {
      expect(getByText(i.toString())).toBeTruthy();
    }
  });

  it("appends the prefix to the links", () => {
    const { getAllByRole } = render(
      <Pagination perPage={6} currentPage={1} total={12} prefix="/products/" />
    );
    expect(
      getAllByRole("link").every((l) =>
        l.getAttribute("href")?.startsWith("/products/")
      )
    ).toBeTruthy();
  });
});
