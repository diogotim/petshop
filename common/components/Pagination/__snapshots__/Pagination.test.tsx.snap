// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders the pagination unchanged 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <a
          class="active"
          href="/products/1"
        >
          1
        </a>
        <a
          class=""
          href="/products/2"
        >
          2
        </a>
        <a
          class=""
          href="/products/3"
        >
          3
        </a>
        <a
          class=""
          href="/products/4"
        >
          4
        </a>
        <a
          class=""
          href="/products/5"
        >
          5
        </a>
        <a
          class=""
          href="/products/6"
        >
          6
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <a
        class="active"
        href="/products/1"
      >
        1
      </a>
      <a
        class=""
        href="/products/2"
      >
        2
      </a>
      <a
        class=""
        href="/products/3"
      >
        3
      </a>
      <a
        class=""
        href="/products/4"
      >
        4
      </a>
      <a
        class=""
        href="/products/5"
      >
        5
      </a>
      <a
        class=""
        href="/products/6"
      >
        6
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
